source/opt/dead_branch_elim_pass.cpp

```diff
@@ -73,260 +72,215 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
+        if (!liveBlocks.count(contBlock)) {
+          unreachableContinues[contBlock] = block;
+        }
+      }
     }
+  }
 
-    // If last block reached, look for next dead branch
-    if (dbi == structuredOrder.end()) continue;
+  for (auto& block : *func) {
+    if (liveBlocks.count(&block)) {
+      // Fix phis so that only live (or unremovable) incoming edges are
+      // present. If the block now only has a single live incoming edge, remove
+      // the phi and replace its uses with its data input.
+      for (auto iter = block.begin(); iter != block.end();) {
+        if (iter->opcode() != SpvOpPhi) {
+          break;
+        }
 
-    // Create set of dead predecessors in preparation for phi update.
-    // Add the header block if the live branch is not the merge block.
-    std::unordered_set<ir::BasicBlock*> deadPreds(elimBlocks);
-    if (liveLabId != dLabId) deadPreds.insert(*bi);
+        bool changed = false;
+        ir::Instruction* inst = &*iter;
+        std::vector<ir::Operand> operands;
+        // Iterate through the incoming labels and determine which to keep
+        // and/or modify.
+        for (uint32_t i = 1; i < inst->NumInOperands(); i += 2) {
+          ir::BasicBlock* inc =
+              get_parent_block(inst->GetSingleWordInOperand(i));
+          if (unreachableContinues.count(inc)) {
+            // Replace incoming value with undef.
+            operands.emplace_back(
+                SPV_OPERAND_TYPE_ID,
+                std::initializer_list<uint32_t>{Type2Undef(inst->type_id())});
+            operands.push_back(inst->GetInOperand(i));
+            changed = true;
+          } else if (liveBlocks.count(inc)) {
+            // Keep live incoming edge.
+            operands.push_back(inst->GetInOperand(i - 1));
+            operands.push_back(inst->GetInOperand(i));
+          } else {
+            // Remove incoming edge.
+            changed = true;
+          }
+        }
 
-    // Update phi instructions in terminating block.
-    ir::Instruction* inst = &*(*dbi)->begin();
-    while (inst) {
-      // Skip NoOps, break at end of phis
-      SpvOp op = inst->opcode();
-      if (op == SpvOpNop) {
-        inst = inst->NextNode();
-        continue;
-      }
-      if (op != SpvOpPhi) break;
-      // Count phi's live predecessors with lcnt and remember last one
-      // with lidx.
-      uint32_t lcnt = 0;
-      uint32_t lidx = 0;
-      uint32_t icnt = 0;
-      inst->ForEachInId([&deadPreds, &icnt, &lcnt, &lidx, this](uint32_t* idp) {
-        if (icnt % 2 == 1) {
-          if (deadPreds.find(cfg()->block(*idp)) == deadPreds.end()) {
-            ++lcnt;
-            lidx = icnt - 1;
+        if (changed) {
+          // Replace the phi with either a single value or a rebuilt phi.
+          uint32_t replId;
+          if (operands.size() == 2) {
+            replId = operands[0].words[0];
+          } else {
+            replId = TakeNextId();
```

Comment:
If we're going to delete the old phi we should reuse its result id, at least in principle.
  
That makes it tricky to keep inst-def-use up to date but it should be doable.  You'd have to kill the old instruction then add the new one.

Reply:
Ok I'll do this.